### PR TITLE
Add discord.Guild converter and GuildNotFound error

### PR DESCRIPTION
**Pull request by AbstractUmbra**
## Summary

This adds a `discord.Guild` converter as well as the resulting error if a match is not found.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/converter.py [hunk 1]
```diff
@@ -524,6 +525,33 @@ async def convert(self, ctx, argument):
...
+    """
+
+    async def convert(self, ctx, argument):
+        match = self._get_id_match(argument)
+        result = None
+
+        if match is not None:
+            guild_id = int(match.group(1))
+            result = ctx.bot.get_guild(guild_id)
+            return result
```
Comment (z03h) [inline discord/ext/commands/converter.py:546]: This could return None instead of raising GuildNotFound

Comment (Gobot1234) [reply discord/ext/commands/converter.py:546]: That wouldn't match the behaviour of any of the current converters

Comment (z03h) [reply discord/ext/commands/converter.py:546]: That's what I mean? I didn't change anything and it currently can return None for an wrong id

File: discord/ext/commands/converter.py [hunk 2]
```diff
@@ -524,6 +525,33 @@ async def convert(self, ctx, argument):
         except Exception as exc:
             raise BadInviteArgument() from exc
 
+class GuildConverter(IDConverter):
+    """Converts to a :class:`~discord.Guild`.
+
+    The lookup strategy is as follows (in order):
```
Comment (Rapptz) [inline discord/ext/commands/converter.py:531]: This should probably mention there's no disambiguation if guilds have multiple matching names.

File: discord/ext/commands/converter.py (hunk 1 above)
Comment (Rapptz) [inline discord/ext/commands/converter.py:546]:
```suggestion
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Historically I'm against such a converter but I guess if the shortcomings are documented it's okay, even though I think the use cases for it aren't that big.

**Review by Zomatree — CHANGES_REQUESTED** [review]

File: discord/ext/commands/converter.py [hunk 3]
```diff
@@ -524,6 +525,33 @@ async def convert(self, ctx, argument):
...
+        match = self._get_id_match(argument)
+        result = None
+
+        if match is not None:
+            guild_id = int(match.group(1))
+            result = ctx.bot.get_guild(guild_id)
+            return result
+
+        predicate = lambda g: g.name == argument
+        result = discord.utils.find(predicate, ctx.bot.guilds)
```
[the remark this answers is not in the document]
Comment (bijij) [reply discord/ext/commands/converter.py:549]: Afaik `uitlls.find` was used to be concistent with other convereters.

[the remark this answers is not in the document]
Comment (Zomatree) [reply discord/ext/commands/converter.py:549]: I can only find two other use of `utils.find` where `utils.get` can be used instead (both in `UserConverter`), most other uses of it have more of a "custom" check that cant be done with `utils.get`, having consistency here doesn't really matter anyway.

File: discord/ext/commands/converter.py (hunk 1 above)
Comment (AbstractUmbra) [reply discord/ext/commands/converter.py]: Fixed in a manner similar to the other converters.

File: discord/ext/commands/converter.py [hunk 4]
```diff
@@ -524,6 +525,33 @@ async def convert(self, ctx, argument):
...
+        match = self._get_id_match(argument)
+        result = None
+
+        if match is not None:
+            guild_id = int(match.group(1))
+            result = ctx.bot.get_guild(guild_id)
+
+        if result is None:
+            predicate = lambda g: g.name == argument
+            result = discord.utils.find(predicate, ctx.bot.guilds)
```
Comment (Rapptz) [inline discord/ext/commands/converter.py:549]: `get` is probably better (it is faster than `find` for the single attribute case).

**Review by Rapptz — CHANGES_REQUESTED** [review]
This is missing the documentation changes required in api.rst

Mainly, adding the class to the list of converters and adding the exception to the exception hierarchy and the list of exceptions.

**Review by Rapptz — APPROVED** [review]

**Pull request merged**